src/index.ts

```diff
@@ -80,36 +81,68 @@ export const orderbooksFetcher = new OrderbookFetcher(web3, argv['page-size'], a
 
 /* tslint:disable:no-unused-expression */
 
+router.get('/metrics', (_req, res) => {
+  res.header('Content-Type', 'text/plain; charset=utf-8')
+  res.send(register.metrics())
+})
+
 router.get('/markets/:base-:quote', async (req, res) => {
-  if (!req.query.atoms) {
-    res.sendStatus(HTTP_STATUS_UNIMPLEMENTED)
-    return
+  const { base, quote } = req.params
+  const { atoms } = req.query
+  const hops = getHops(req, argv['max-hops'])
+
+  getOrderBookMetrics.total.inc()
+  getOrderBookMetrics.requests.inc({ base, quote, hops })
+  const endTimerTotal = getOrderBookMetrics.durationsTotals.startTimer()
+  const endTimerMarket = getOrderBookMetrics.durations.startTimer()
+  try {
+    if (!atoms) {
+      res.sendStatus(HTTP_STATUS_UNIMPLEMENTED)
+      return
+    }
+    const serialized = orderbooksFetcher.serializeOrderbooks()
+    const result = await pool.exec('markets', [serialized, base, quote, hops])
+    res.json(result)
+  } catch (error) {
+    getOrderBookMetrics.errors.inc()
+    throw error
+  } finally {
+    endTimerTotal()
+    endTimerMarket({ base, quote, hops })
```

Comment:
I see a pattern here that we could extract into a function:

``` typescript
async function executeWithMetrics(fn, params, metrics) {
  metrics.total.inc()
  metrics.requests.inc(params)
  const endTimerTotal = metrics.durationsTotals.startTimer()
  const endTimerMarket = metrics.durations.startTimer()
  try {
    return fn()
  catch (error) {
    metrics.errors.inc()
  } finally {
    endtimerTotal()
    endTimerMarket(params)
  }
}
```

and on the request handler itself:

``` typescript
router.get('/markets/:base-:quote', async (req, res) => {
  if (!req.query.atoms) {
    res.sendStatus(HTTP_STATUS_UNIMPLEMENTED)
    return
  const { base, quote } = req.params
  const { atoms } = req.query
  const hops = getHops(req, argv['max-hops'])

  return executeWithMetrics(() => {
    if (!atoms) {
      res.sendStatus(HTTP_STATUS_UNIMPLEMENTED)
      return
    }
    const serialized = orderbooksFetcher.serializeOrderbooks()
    const result = await pool.exec('markets', [serialized, base, quote, hops])
    res.json(result)
  }, {base, quote, hops}, getOrderBookMetrics)

```

Reply:
Implemented refactor here:
https://github.com/gnosis/dex-price-estimator/pull/26/commits/32c90cf9eab5b49940bd900edec710fbcc875e10

Reply:
:heart_eyes: